libkineto/src/output_json.cpp

```diff
@@ -257,22 +261,56 @@ void ChromeTraceLogger::handleGenericActivity(
       fmt::format("{}", op.resourceId());
 
   // clang-format off
-  traceOf_ << fmt::format(R"JSON(
-  {{
-    "ph": "X", "cat": "{}", {},
-    "args": {{
-       "External id": {},
-       "Trace name": "{}", "Trace iteration": {}{}
-       {}
-    }}
-  }},)JSON",
-      toString(op.type()), traceActivityJson(op),
-      // args
-      op.id,
-      op.traceSpan()->name, op.traceSpan()->iteration, separator,
-      op_metadata);
-  // clang-format on
 
+  switch (op.type()) {
+    case ActivityType::CUDA_RUNTIME:
```

Comment:
OK this is pretty hacky, but clearly GenericActivity is not generic enough yet.
Perhaps we can live with this for a bit but let's create an issue for fixing GenericTraceActivity.

Reply:
Well, it has some endearing qualities:

1. it's in the right place (file), this is a json formatting concern.
2. I didn't have to make several subclasses of TraceActivity
3. I didn't have to bloat output_json.h with methods to pass those in
4. I didn't have ifdef those methods to keep the interface generic
5. The choice of metadata to output (a gpu concern) is made in RoctracerActivityInterface (the right place) rather than in output.json which shouldn't really shouldn't be privy to those details.

End of the day this switches off TraceActivity.type, while the cupti path switches of of typeid.  Six of one... half dozen of the other.

TraceActivity defines the ActivityType so it acknowledges there are different types of activity.  This function deals with only those types, that seems like fair game.
 
Granted this is a tricky pattern to get right from an OO and SOC point of view.  We'll figure something nice out.

Reply:
Ah, I definitely think using the generic activity makes sense, I was just talking about the switch statement. I think we can get rid of that with some small changes.